api/catalog/api/controllers/elasticsearch/utils.py

```diff
@@ -55,26 +64,71 @@ def paginate_with_dead_link_mask(
     :param page: The page number.
     :return: Tuple of start and end.
     """
+    logger = parent_logger.getChild("paginate_with_dead_link_mask")
     query_hash = get_query_hash(s)
     query_mask = get_query_mask(query_hash)
+    logger.debug(
+        f"page_size={page_size} "
+        f"page={page} "
+        f"query_hash={query_hash} "
+        f'query_mask={",".join(map(str, query_mask))} '
+    )
     if not query_mask:
         start = 0
         end = ceil(page_size * page / (1 - DEAD_LINK_RATIO))
-    elif page_size * (page - 1) > sum(query_mask):
+        logger.debug(
+            "empty query_mask "
+            f"page_size={page_size} "
+            f"page={page} "
+            f"start={start} "
+            f"end={end} "
+        )
+        return start, end
+
+    # TODO: What does this request_size mean vs the one below
+    # that doesn't subtract 1 from the page? Why do we subtract
+    # 1 here and the other time we do not?
```

Comment:
The refactoring here looks good to me in that it preserves the original logic, but I agree that I am also confused 😄 Excited to have the logs to allow for investigation.